src/main/java/org/elasticsearch/search/aggregations/bucket/BestDocsDeferringCollector.java

```diff
@@ -113,50 +115,110 @@ public void preCollection() throws IOException {
 
     @Override
     public void postCollection() throws IOException {
-        finished = true;
+        runDeferredAggs();
     }
 
-    /**
-     * Replay the wrapped collector, but only on a selection of buckets.
-     */
+
     @Override
     public void prepareSelectedBuckets(long... selectedBuckets) throws IOException {
-        if (!finished) {
-            throw new ElasticsearchIllegalStateException("Cannot replay yet, collection is not finished: postCollect() has not been called");
-        }
-        if (selectedBuckets.length > 1) {
-            throw new ElasticsearchIllegalStateException("Collection only supported on a single bucket");
-        }
+        // Historically we run the deferred aggs here but a test with
+        // terms->sampler->max aggs was failing if the top-level terms agg had
+        // an Order setting that referred to the leaf Max value. This is because
+        // before calling this method the terms agg would try access the value
+        // of the Max agg which was unset because there had been no collections.
+        // That is why the runDeferredAggs logic was moved to the postCollection
+        // call above.
```

Comment:
This comment is useful in the context of this review to know why the code moved but given that this was new as of a couple of weeks ago I think it can be removed before pushing?